src/rewriting/composition.rs

```diff
@@ -0,0 +1,379 @@
+        let right_new_name = right.name().to_string();
+
+        // Preserve the schema names of the existing JOIN
+        let schema_names: Vec<String> =
+            join.schema().iter().map(|f| f.name().to_string()).collect();
+        let left_names: Vec<String> = schema_names
+            .iter()
+            .take(join.left().schema().len())
+            .cloned()
+            .collect();
+        let right_names: Vec<String> = schema_names
+            .iter()
+            .skip(join.left().schema().len())
+            .cloned()
+            .collect();
+
+        let columns_mapping: Hierarchy<Identifier> = join
+            .left()
+            .schema()
+            .iter()
+            .zip(left.schema().iter())
+            .map(|(o, n)| {
+                (
+                    vec![join.left().name().to_string(), o.name().to_string()],
+                    Identifier::from(vec![left_new_name.clone(), n.name().to_string()]),
+                )
+            })
+            .chain(
+                join.right()
+                    .schema()
+                    .iter()
+                    .zip(right.schema().iter())
+                    .map(|(o, n)| {
+                        (
+                            vec![join.right().name().to_string(), o.name().to_string()],
+                            Identifier::from(vec![right_new_name.clone(), n.name().to_string()]),
+                        )
+                    }),
+            )
+            .collect();
+
+        // build the output relation
+        Relation::join()
+            .left_names(left_names)
+            .right_names(right_names)
+            .operator(join.operator().clone().rename(&columns_mapping))
+            .left(left)
+            .right(right)
+            .build()
+    }
+
+    fn set(&self, set: &'a Set, left: Relation, right: Relation) -> Relation {
+        Relation::set().with(set.clone()).left(left).right(right).build()
+    }
+
+    fn values(&self, values: &'a Values) -> Relation {
+        Relation::Values(values.clone())
+    }
+}
+
+
+fn composer_visitor<'a>(
+    relations: &'a Hierarchy<Arc<Relation>>,
+) -> ComposerVisitor<impl Fn(&Table) -> Relation + 'a> {
+    ComposerVisitor::new(move |table: &Table| {
+        match relations.get(table.path()) {
+            Some(r) => r.deref().clone(),
+            None => table.clone().into()
+        }
+    })
+}
+
+impl Relation {
+    pub fn compose<'a>(
+        &'a self,
+        relations: &'a Hierarchy<Arc<Relation>>,
+    ) -> Relation {
+        self.accept(composer_visitor(relations))
+    }
+}
+
+/// Function used to compose relations. It substitutes Tables in the
+/// outer_relations with relations in inner_relations with the same path.
+/// The output relations will have the same paths as the outer_relations
+/// Tables in the outer_relations should have a compatible schema with
+/// inner_relations otherwise an error is raised.
+pub fn compose_relations<'a>(
```

Comment:
Why not make this a method of `Hierarchy<Arc<Relation>>`?

Can't you implement this
```
impl Hierarchy<Arc<Relation>> {
  pub fn compose()...
}
```
in this module ?

Reply:
ok, Implemented  `Hierarchy<Arc<Relation>>` with the `compose` method.